### ISSUE TITLE
streaming error

### ISSUE THREAD

**Issue by wisechengyi** [gradle/ge-export#2]
Hi @w25r ,

Thanks for the help on initial configuration, and I was finally able to create a local db and initiate ge-export.

Here is the output I have while running. Question:
1) is this supposed to one time process where it gets the last 24 hours of data and finish, or 
2) get the last 24 hours of data then keep streaming? in which case, I am not sure if something is wrong here, or is it expected to show the streaming error if there is no incoming builds.
```
$ ./gradlew run -Dserver=https://<ge_hostname> -Dhours=24 -DcreateDb
Starting a Gradle Daemon (subsequent builds will be faster)
:compileJava UP-TO-DATE
:processResources
:classes
:run
13:21:27,657 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Could NOT find resource [logback-test.xml]
13:21:27,657 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Could NOT find resource [logback.groovy]
13:21:27,658 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Found resource [logback.xml] at [file:/Users/yic/workspace/ge-export/build/resources/main/logback.xml]
13:21:27,746 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - debug attribute not set
13:21:27,746 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [ch.qos.logback.core.ConsoleAppender]
13:21:27,754 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [STDOUT]
13:21:27,812 |-WARN in ch.qos.logback.core.ConsoleAppender[STDOUT] - This appender no longer admits a layout as a sub-component, set an encoder instead.
13:21:27,812 |-WARN in ch.qos.logback.core.ConsoleAppender[STDOUT] - To ensure compatibility, wrapping your layout in LayoutWrappingEncoder.
13:21:27,812 |-WARN in ch.qos.logback.core.ConsoleAppender[STDOUT] - See also http://logback.qos.ch/codes.html#layoutInsteadOfEncoder for details
13:21:27,812 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [ch.qos.logback.core.FileAppender]
13:21:27,814 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [FILE]
13:21:27,816 |-INFO in ch.qos.logback.core.joran.action.NestedComplexPropertyIA - Assuming default type [ch.qos.logback.classic.encoder.PatternLayoutEncoder] for [encoder] property
13:21:27,817 |-INFO in ch.qos.logback.core.FileAppender[FILE] - File property is set to [export.log]
13:21:27,819 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [com.base22] to TRACE
13:21:27,819 |-INFO in ch.qos.logback.classic.joran.action.RootLoggerAction - Setting level of ROOT logger to DEBUG
13:21:27,819 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [FILE] to Logger[ROOT]
13:21:27,820 |-INFO in ch.qos.logback.classic.joran.action.RootLoggerAction - Setting level of ROOT logger to INFO
13:21:27,820 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [STDOUT] to Logger[ROOT]
13:21:27,820 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - End of configuration.
13:21:27,821 |-INFO in ch.qos.logback.classic.joran.JoranConfigurator@de0a01f - Registering current configuration as safe fallback point

13:21:28.219 [main] INFO  com.zaxxer.hikari.HikariDataSource - yank-default - Started.
13:21:28.349 [main] INFO  com.zaxxer.hikari.pool.PoolBase - yank-default - Driver does not support get/set network timeout for connections. (Method org.postgresql.jdbc.PgConnection.getNetworkTimeout() is not yet implemented.)
13:21:28.369 [main] INFO  org.knowm.yank.YankPoolManager - Initialized pool 'yank-default'
13:21:28.370 [main] INFO  com.gradle.exportapi.CreateDB - Creating Database
13:21:28.789 [main] INFO  com.gradle.exportapi.Application - lastBuildEventId: null
13:21:28.789 [main] INFO  com.gradle.exportapi.Application - Builds uri: /build-export/v1/builds/since/1496175688773
13:21:28.835 [main] INFO  com.gradle.exportapi.Application - Streaming builds...
13:21:29.344 [rxnetty-nio-eventloop-1-1] INFO  com.gradle.exportapi.Application - Error streaming builds, resuming from build id: null
13:21:29.345 [rxnetty-nio-eventloop-1-1] INFO  com.gradle.exportapi.Application - Builds uri: /build-export/v1/builds/since/1496175688773
13:21:29.345 [rxnetty-nio-eventloop-1-1] INFO  com.gradle.exportapi.Application - Streaming builds...
13:21:29.393 [rxnetty-nio-eventloop-1-1] INFO  com.gradle.exportapi.Application - Error streaming builds, resuming from build id: null
13:21:29.393 [rxnetty-nio-eventloop-1-1] INFO  com.gradle.exportapi.Application - Builds uri: /build-export/v1/builds/since/1496175688773
13:21:29.394 [rxnetty-nio-eventloop-1-1] INFO  com.gradle.exportapi.Application - Streaming builds...
13:21:29.434 [rxnetty-nio-eventloop-1-1] INFO  com.gradle.exportapi.Application - Error streaming builds, resuming from build id: null
13:21:29.435 [rxnetty-nio-eventloop-1-1] INFO  com.gradle.exportapi.Application - Builds uri: /build-export/v1/builds/since/1496175688773
13:21:29.435 [rxnetty-nio-eventloop-1-1] INFO  com.gradle.exportapi.Application - Streaming builds...
13:21:29.479 [rxnetty-nio-eventloop-1-1] INFO  com.gradle.exportapi.Application - Error streaming builds, resuming from build id: null
13:21:29.479 [rxnetty-nio-eventloop-1-1] INFO  com.gradle.exportapi.Application - Builds uri: /build-export/v1/builds/since/1496175688773
.... (similar messages keep going)
```

I'd guess something is wrong since my destination db is empty
```
postgres=# \dt
             List of relations
 Schema |     Name      | Type  |  Owner   
--------+---------------+-------+----------
 public | builds        | table | postgres
 public | custom_values | table | postgres
 public | tasks         | table | postgres
 public | tests         | table | postgres
(4 rows)

postgres=# 
postgres=# select * from builds;
 id | build_id | user_name | root_project_name | start | finish | status | tags 
----+----------+-----------+-------------------+-------+--------+--------+------
(0 rows)

postgres=# select * from tasks;
 id | build_id | task_id | path | type | duration_millis | outcome 
----+----------+---------+------+------+-----------------+---------
(0 rows)

postgres=# select * from tests;
 id | build_id | task_id | test_id | name | class_name | status | duration_millis 
----+----------+---------+---------+------+------------+--------+-----------------
(0 rows)

```

Also I noticed that my cpu is around 200% doing this.

**Comment by w25r** [comment]
It appears it's not properly resolving the server URL... I'll need to add some verification to the URL for sure.  

A couple of things to try:  Is your server URL correct?  Is it password protected (if so you'll need to provide the password)?  Is it running on the default HTTPS port?

**Comment by wisechengyi** [comment]
i see. it was able to work with `-Dbasic_auth=user:pwd` in addition to the original command.

no password is required to see the regular GUI so i thought no auth would be needed. but yeah, it would be good to clarify admin credential is needed for this one.

**Comment by wisechengyi** [comment]
although it is a bit sketchy that cred is passed on cli..

**Comment by w25r** [comment]
The credentials can be encoded, but yes, still a bit sketchy.

**Comment by wisechengyi** [comment]
Hi @w25r, 

I was wondering if you would be able to help on the following questions:

1. E.g. given `-Dhours=24`, it will grab the data for last 24 hours, then keep on streaming forever for new builds, correct?

2. Currently we do not support postgres internally, as in I would have to maintain any postgres db, whereas mysql has a dedicated team so I only need to think about how to use it. Do you think it would be easy to export the data in mysql format?

Thanks,
Yi

**Comment by w25r** [comment]
The tool as written will not stream forever.  It could be adopted to do just that, and a workaround in the meantime would be to run a cronjob every few hours and export the last few hours of builds.  

I have logged #3 for your second question.  Any chance you could submit a pull request for MySQL support?  If not, I will try to get to it soon.

**Comment by wisechengyi** [comment]
Thanks @w25r . I'll be happy to submit incremental changes down the line, but given that I am still exploring the overall scheme how to integrate gradle enterprise internally, it would probably take me a longer to get to #3, so your help is definitely appreciated!

**Comment by wisechengyi** [comment]
re having a cron job to export periodically:

1. `./gradlew run -Dserver=https://<my_server> -Dhours=all -DcreateDb` to create initiate db
2. Purposely ctrl+c out of it before it finishes
3. `./gradlew run -Dserver=https://<my_server> -Dhours=all` (without `-DcreateDb`) this time, then got the following error complain the build id already exists
```
13:03:34.484 [rxnetty-nio-eventloop-1-1] INFO  com.gradle.exportapi.Application - Streaming events for : rf5vofq6lonos
13:03:34.497 [main] ERROR org.knowm.yank.Yank - Error in SQL query!!!; ERROR: duplicate key value violates unique constraint "unique_build_id"
  Detail: Key (build_id)=(xwl2azyp45l7w) already exists. Query: INSERT INTO builds (build_id, user_name, root_project_name, start, finish, status, tags) VALUES (?,?,?,?,?,?,?) Parameters: [xwl2azyp45l7w, yic, twitter-android, 2017-05-16T22:47:52.030-07:00, 2017-05-16T22:47:52.030-07:00, finished, ]; Pool Name= yank-default; SQL= INSERT INTO builds (build_id, user_name, root_project_name, start, finish, status, tags) VALUES (?,?,?,?,?,?,?)
org.knowm.yank.exceptions.YankSQLException: Error in SQL query!!!; ERROR: duplicate key value violates unique constraint "unique_build_id"
  Detail: Key (build_id)=(xwl2azyp45l7w) already exists. Query: INSERT INTO builds (build_id, user_name, root_project_name, start, finish, status, tags) VALUES (?,?,?,?,?,?,?) Parameters: [xwl2azyp45l7w, yic, twitter-android, 2017-05-16T22:47:52.030-07:00, 2017-05-16T22:47:52.030-07:00, finished, ]; Pool Name= yank-default; SQL= INSERT INTO builds (build_id, user_name, root_project_name, start, finish, status, tags) VALUES (?,?,?,?,?,?,?)
        at org.knowm.yank.Yank.handleSQLException(Yank.java:676)
        at org.knowm.yank.Yank.insert(Yank.java:127)
        at org.knowm.yank.Yank.insert(Yank.java:107)
        at com.gradle.exportapi.dao.BuildDAO.insertBuild(BuildDAO.java:36)
        at com.gradle.exportapi.EventProcessor.persist(EventProcessor.java:221)
        at rx.observables.BlockingObservable$9.onNext(BlockingObservable.java:669)
        at rx.internal.operators.NotificationLite.accept(NotificationLite.java:135)
        at rx.observables.BlockingObservable.subscribe(BlockingObservable.java:534)
        at rx.observables.BlockingObservable.subscribe(BlockingObservable.java:666)
```
So the question is, should `ge-export` be smart enough to figure out the difference? otherwise the cron job will fail for any tiny overlap between the exports.

**Comment by w25r** [comment]
I agree and have logged https://github.com/gradle/ge-export/issues/4

**Comment by w25r** [comment]
I think the various issues have been logged separately:

Keep alive is in #5 
MySQL support is in #3 
Duplicate records is in #4 

And much of the troubleshooting is being fixed in #10.

I will close this issue and address them individually.
